### PR TITLE
Fix unsound equivalence between Dates and numbers

### LINKED ISSUE #5539: Major: Function type with Date argument accepts a function with number argument
### Flow verisons
Seems to happen on any version I've tried so far:
* 0.37
* 0.40
* 0.42
* 0.50
* 0.57.1
* 0.61

### Code [(Try flow link)](https://flow.org/try/#0C4TwDgpgBAIghsCBhOAbVAjOBjA1lAXigAoA3NAVwgC5YEIBKQgPigAUAnAewFsBLAM4QAPHAB2IZgCgoUKQDMKY7MD5cxUeVy5lKNKGIo8MEDg1qdegkeMlQA3jKgcIwChw2X+QgHQuBXKikEMQMUgC+UlLEWly08Igo6Fh4DADcQA)
```js
type DateCallback = (value: Date) => Promise<any>
  
function foo(value: number): Promise<any> {
  return Promise.resolve()
}

(foo: DateCallback);
```

### Expected
```
7: (foo: DateCallback);
     ^ function. This type is incompatible with
7: (foo: DateCallback);
          ^ function type
This parameter is incompatible:
1: type DateCallback = (value: Date) => Promise<any>
                               ^ Date. This type is incompatible with
3: function foo(value: number): Promise<any> {
                       ^ number
```

### Actual
```
No errors!
```

### PR DESCRIPTION
**Pull request by NilSet**
fixes #5539

### CODE REVIEW

File: tests/date/date.js [hunk 1]
```diff
@@ -24,3 +24,9 @@ new Date(2015, 6, 18, 11, 55, '42');
 new Date(2015, 6, 18, 11, 55, 42, '999');
 new Date(2015, 6, 18, 11, 55, 42, 999, 'hahaha');
 new Date('2015', 6);
+
+var b:boolean = d > 0;
+var n1:number = d - 12;
+var n2:number = d & 255;
+var n3:number = d + 12; // expect error
+var s:string = d + 12; // fixme? in js this coerces both to string and concats them
```
Comment (ilias-t) [inline tests/date/date.js]: Assuming the maybe FIXME is that this should throw an error, but does not. Should we file a separate issue for this & reference in the comment?

Comment (NilSet) [reply tests/date/date.js]: It throws confusing errors, and it possibly should be allowed.

File: src/typing/flow_js.ml [hunk 2]
```diff
@@ -7308,12 +7308,15 @@ and any_propagated_use cx trace use_op any = function
 and numeric = function
   | DefT (_, NumT _) -> true
   | DefT (_, SingletonNumT _) -> true
+  | _ -> false
 
+and dateiform = function
```
[the remark this answers is not in the document]
Comment (ilias-t) [reply src/typing/flow_js.ml]: Ok so -iform is like -esque 👍

**Comment by mroch** [comment]
I'm trying to remember why we went out of our way to support this. I'm going to import it to test on our internal repositories to see if that surfaces the reason.

**Comment by NilSet** [comment]
Based on the discussion in #5539 it may have been an unintended consequence of allowing numeric operations and comparisons on Dates. This PR preserves that quality of life feature, and still maintains a type level distinction between Date and number, which is important since it very much is not a number.

An example of the unsoundness is
```
const a: number = new Date() + 12;
```
leaving a string tagged in the type system as a number.

**Comment by NilSet** [comment]
I think at least some of this problem was introduced in a7cb6da61d26f8590272571c1f8ca2ad762b08ba

**Comment by mroch** [comment]
thanks! one common pattern we used was `new Date(date)` to clone a date, so I added a constructor for that. otherwise, whatever motivation we had wasn't worth it because this PR uncovered a bunch of legit bugs.

**Comment by goodmind** [comment]
But this still allows multiplying `Date`?
https://flow.org/try/#0MYewdgzgLgBAhgLhmArgWwEYFMBOMC8yWA7jACJxRYAUAlDAFQwCMATANxA

**Pull request closed without merging**